service/controller/resource/appfinalizer/delete.go

```diff
@@ -0,0 +1,92 @@
+package appfinalizer
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/controller/context/finalizerskeptcontext"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/giantswarm/cluster-operator/pkg/label"
+)
+
+// EnsureDeleted removes finalizers for workload cluster app CRs. These are
+// deleted with the cluster by the provider operator.
+func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
+	clusterConfig, err := r.getClusterConfigFunc(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// We keep the finalizer for the app-operator app CR so the resources in
+	// the management cluster are deleted.
+	o := metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s!=%s", label.AppKubernetesName, "app-operator"),
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "finding apps to remove finalizers for")
+
+	list, err := r.g8sClient.ApplicationV1alpha1().Apps(clusterConfig.ID).List(o)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d apps to remove finalizers for", len(list.Items)))
+
+	for _, app := range list.Items {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("removing finalizer for app %#q", app.Name))
+
+		index := getFinalizerIndex(app.Finalizers)
+		if index >= 0 {
+			patches := []patch{
+				{
+					Op:   "remove",
+					Path: fmt.Sprintf("/metadata/finalizers/%d", index),
```

Comment:
Can we simply remove all the finalizers?

Reply:
I'd prefer to keep it the same as the master branch.

Also app-operator has the migration logic to set the finalizer to the one being removed here. So I think it's safer to just remove that finalizer.